include/rocksdb/slice.h

```diff
@@ -195,8 +195,15 @@ class PinnableSlice : public Slice, public Cleanable {
     }
   }
 
-  void remove_prefix(size_t /*n*/) {
-    assert(0);  // Not implemented
+  void remove_prefix(size_t n) {
+    assert(n <= size());
+    if (pinned_) {
+      data_ += n;
+      size_ -= n;
+    } else {
+      buf_->erase(0, n);
```

Comment:
Is it safe to do this if buf was passed to the constructor and thus not owned by the PinnableSlice? Also, it might be inefficient. Why not just change data_ and size_ in this case too?

Reply:
The logic was to keep data_, and size_ consistent with what buf_ represents. There is a GetSelf function that returns the buf_ and if what they get from GetSelf is different than what what data_ and size_ indicate that might cause bugs in future. It does not have to be like that and we could have had defined the contract of GetSelf not to have to be exactly what data() and size() indicate. But the existing implementation of remove_suffix does respect this consistency, so it is more sensible to follow the pattern.

About efficiency, I am not concerned about existing usages as there is none. And we dont expect anyone to use it either. One might say we can postpone efficiency/clarity tradeoff decision for future, and update both remove_prefix and remove_suffix implementations.

About safety, it follows what we already have in remove_suffix implementation. So I dont expect the users to be surprised about what happens to the buffer that they pass to PinnableSlice.

Reply:
Its a good point about GetSelf(). It is a problematic function and requires callers to know whether the slice is pinned or not. So we can follow the pattern of remove_suffix() and revisit this at some point.